### PR TITLE
Updating att_dis to support multiple attributes

### PR DESCRIPTION
**Pull request by ryanzhang22**

### CODE REVIEW

File: analysis_notebooks/Attribute Analysis.ipynb [hunk 1]
```diff
@@ -981,6 +984,296 @@
...
+    "\n",
+    "        cat_p_values.append(p)\n",
+    "        cat_distr_dists.append([a_dist, b_dist])\n",
+    "        cat_dists.append([np.mean(a_dist), np.mean(b_dist)])\n",
+    "        cat_file_paths.append([a_file, b_file])\n",
+    "        cat_indices.append([a_index, b_index])\n",
+    "\n",
+    "        a_mean = np.mean(a_dist)\n",
+    "        b_mean = np.mean(b_dist)\n",
+    "        if not np.isnan(a_mean) and not np.isnan(b_mean) and not len(a_dist) < 5 and not len(b_dist) < 5:\n",
```
Comment (Angelina-Wang) [inline analysis_notebooks/Attribute Analysis.ipynb]: maybe len(a_dist) > 5 is more readable than not <?

File: analysis_notebooks/Attribute Analysis.ipynb [hunk 2]
```diff
@@ -981,6 +984,296 @@
...
+    "        bpb = plt.boxplot(b, sym='', widths=0.6)\n",
+    "        plt.close()\n",
+    "\n",
+    "        save_box_images(bpa, a, attr_file_paths[cat_index][0], attr_indices[cat_index][0], \n",
+    "            bpb, b, attr_file_paths[cat_index][1], attr_indices[cat_index][1],\n",
+    "            xaxis_sig[cat_index], cat_index, to_save)\n",
+    "    except (AttributeError, FileNotFoundError): \n",
+    "        print('Some functionality not available for CoCoDatasetNoImages Class')"
+   ]
+  },
```
Comment (Angelina-Wang) [inline analysis_notebooks/Attribute Analysis.ipynb:1227]: in the format of these notebooks, each metric has a "setup" header and then an "analyses" header, this seems to be missing the latter

File: measurement_files/attribute_based.py [hunk 3]
```diff
@@ -211,12 +241,12 @@ def att_clu(dataloader, args):
...
                 small_data = F.interpolate(instance.unsqueeze(0), size=32, mode='bilinear').to(device)
                 this_small_features = small_model.features(small_data)
-                if len(scene_features[index][attr[0]]) < 500 and index not in scene_added:
+                if len(scene_features[index][attr[0]-1]) < 500 and index not in scene_added:
                     scene_added.append(index)
-                    scene_features[index][attr[0]].extend(this_features.data.cpu().numpy())
-                    scene_filepaths[index][attr[0]].append((target[3], pred))
-                if len(instance_features[index][attr[0]]) < 500:
-                    instance_features[index][attr[0]].extend(this_small_features.data.cpu().numpy())
+                    scene_features[index][attr[0]-1].extend(this_features.data.cpu().numpy())
```
Comment (Angelina-Wang) [inline measurement_files/attribute_based.py]: why is this "-1" there?

File: measurement_files/attribute_based.py [hunk 4]
```diff
@@ -153,6 +153,36 @@ def count_cooccurrence(dataloader, args):
...
+                distance = np.linalg.norm(person_center - ann_center)
+                # Finds the person-object pair that has the shortest distance
+                if categories.index(ann['label']) in seen_instances:
+                    prev = distances[categories.index(ann['label'])][attr[0] - 1][-1]
+                    prev_calc_dist = prev[0] / np.sqrt(prev[1]*prev[2])
+                    calc_dist = distance / np.sqrt(person_area*ann_area)
+                    if calc_dist < prev_calc_dist:
+                        distances[categories.index(ann['label'])][attr[0] - 1][-1] = (distance, person_area, ann_area, file_path, j)
+                else:
+                    distances[categories.index(ann['label'])][attr[0] - 1].append((distance, person_area, ann_area, file_path, j))
```
Comment (Angelina-Wang) [inline measurement_files/attribute_based.py:181]: same here, why's there a -1 all over?

File: measurement_files/attribute_based.py (hunk 3 above)
Comment (Angelina-Wang) [reply measurement_files/attribute_based.py]: also throughout this file

File: analysis_notebooks/Attribute Analysis.ipynb [hunk 5]
```diff
@@ -981,6 +984,296 @@
...
+    "\n",
+    "all_ratios = []\n",
+    "for i in range(len(attr_sig_dists)):\n",
+    "    in_count = attr_sig_dists[i][0]\n",
+    "    out_count = attr_sig_dists[i][1]\n",
+    "    if in_count > out_count:\n",
+    "        all_ratios.append(in_count/out_count)\n",
+    "    else:\n",
+    "        all_ratios.append(out_count/-in_count)\n",
+    "    if np.absolute(all_ratios[-1]) <= 1:\n",
```
Comment (Angelina-Wang) [inline analysis_notebooks/Attribute Analysis.ipynb]: what's this here for? if the ratio is 1?

File: analysis_notebooks/Attribute Analysis.ipynb [hunk 6]
```diff
@@ -981,6 +984,296 @@
...
+    "        cat_indices.append([a_index, b_index])\n",
+    "\n",
+    "        a_mean = np.mean(a_dist)\n",
+    "        b_mean = np.mean(b_dist)\n",
+    "        if not np.isnan(a_mean) and not np.isnan(b_mean) and not len(a_dist) < 5 and not len(b_dist) < 5:\n",
+    "            if (a_mean >= b_mean and a_mean/b_mean > largest_ratio and p < 0.05):\n",
+    "                largest_ratio = a_mean/b_mean\n",
+    "                ratio_idx = j\n",
+    "            if (a_mean < b_mean and b_mean/a_mean > largest_ratio and p < 0.05):\n",
+    "                largest_ratio = b_mean/a_mean\n",
```
Comment (Angelina-Wang) [inline analysis_notebooks/Attribute Analysis.ipynb]: does this mean that for one object, only one ratio can be shown? what if group 1 is super different than the rest, but group 3 is also super different than the rest for the same object. both should be able to be shown for this object

File: analysis_notebooks/Attribute Analysis.ipynb (hunk 5 above)
Comment (ryanzhang22) [reply analysis_notebooks/Attribute Analysis.ipynb]: Is this referencing line 1096? That's also there in the gender analysis notebook, I assumed it was there to prune the ratios that were small, but now that I think about it it probably isn't needed.
